### PR TITLE
fix for x509 store add crl

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
ZD 10430

### CODE REVIEW

**Comment by JacobBarthelmeh** [comment]
Fixes two reports. Issue of if the callback has not been set with CRL and issue with wolfSSL_X509_STORE_add_crl not using the CRL added when verifying certificate.

File: src/crl.c [hunk 1]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+        current = current->next;
+    }
+    return head;
+}
+
+
+/* Duplicates everything except the parent cm pointed to.
+ * Expects that Init has already been done to 'dup'
+ * return 0 on success */
+static int DupX509_CRL(WOLFSSL_X509_CRL *dup, WOLFSSL_X509_CRL* crl)
```
Comment (dgarske) [inline src/crl.c:620]: Should the `crl` arg be `const` here?

File: src/crl.c [hunk 2]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+                head = tmp;
+        }
+        current = current->next;
+    }
+    return head;
+}
+
+
+/* returns a deep copy of ent on success and null on fail */
+static CRL_Entry* DupCRL_Entry(CRL_Entry* ent, void* heap)
```
Comment (dgarske) [inline src/crl.c:537]: Can `ent` be `const`?

File: src/crl.c [hunk 3]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+
+
+/* returns head of copied list that was alloc'd */
+static RevokedCert *DupRevokedCertList(RevokedCert* in, void* heap)
+{
+    RevokedCert* head = NULL;
+    RevokedCert* current = in;
+    RevokedCert* prev = NULL;
+    while (current) {
+        RevokedCert* tmp = (RevokedCert*)XMALLOC(sizeof(RevokedCert), heap,
```
Comment (dgarske) [inline src/crl.c:518]: Does it make sense to silently fail memory errors here?

File: src/crl.c [hunk 4]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+    dup->verified = ent->verified;
+
+    if (!ent->verified) {
+        dup->tbsSz = ent->tbsSz;
+        dup->signatureSz = ent->signatureSz;
+        dup->signatureOID = ent->signatureOID;
+        dup->toBeSigned = (byte*)XMALLOC(dup->tbsSz, heap,
+                                          DYNAMIC_TYPE_CRL_ENTRY);
+        if (dup->toBeSigned == NULL) {
+            XFREE(dup, heap, DYNAMIC_TYPE_CRL_ENTRY);
```
Comment (dgarske) [inline src/crl.c:564]: Fail case would leak `dup->certs`. Is there a CRL cleanup function that can be used?

File: src/crl.c [hunk 5]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
+            return NULL;
+        }
+
+        dup->signature = (byte*)XMALLOC(dup->signatureSz, heap,
+                                         DYNAMIC_TYPE_CRL_ENTRY);
+        if (dup->signature == NULL) {
+            XFREE(dup, heap, DYNAMIC_TYPE_CRL_ENTRY);
```
Comment (dgarske) [inline src/crl.c:571]: Same... `dup->certs` list would be leaked on mem failure here.

File: src/crl.c [hunk 6]
```diff
@@ -490,27 +490,204 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
...
     WOLFSSL_ENTER("wolfSSL_X509_STORE_add_crl");
-    if (store == NULL || newcrl == NULL)
+    if (store == NULL || newcrl == NULL || store->cm == NULL)
         return BAD_FUNC_ARG;
 
-    crl = store->crl;
-    crle = newcrl->crlList;
+    if (store->cm->crl == NULL) {
+        crl = wolfSSL_X509_crl_new(store->cm);
+        DupX509_CRL(crl, newcrl);
```
Comment (dgarske) [inline src/crl.c:663]: Missing check of the `DupX509_CRL` return code.

**Comment by JacobBarthelmeh** [comment]
Thanks for the review David. Added in the fixes and a test case.

**Pull request merged**